sources/engine/Xenko.Voxels/Voxels/Voxelization/VoxelStorage/VoxelStorageClipmaps.cs

```diff
@@ -0,0 +1,368 @@
+                    BaseVoxelMatrix = BaseVoxelMatrix * Matrix.Translation(-0.5f,-0.5f,-0.5f);
+                    BaseVoxelMatrix = BaseVoxelMatrix * Matrix.Scaling(2f, 2f, 2f);
+                }
+                */
+                VoxelStorerClipmap Storer = new VoxelStorerClipmap
+                {
+                    storageUints = storageUints,
+                    FragmentsBuffer = FragmentsBuffer,
+                    ClipMapCount = ClipMapCount,
+                    ClipMapCurrent = ClipMapCurrent,
+                    ClipMapResolution = ClipMapResolution,
+                    PerMapOffsetScale = PerMapOffsetScale,
+                    UpdatesPerFrame = UpdatesPerFrame
+                };
+                foreach (var attr in data.Attributes)
+                {
+                    attr.Attribute.CollectVoxelizationPasses(data.passList, Storer, BaseVoxelMatrix, ClipMapResolution, attr.Stage, attr.Output);
+                }
+            }
+            if (UpdatesPerFrame == UpdateMethods.AllClipmapsMultipleRenders)
+            {
+                for (int i = 0; i < ClipMapCount; i++)
+                {
+                    VoxelStorerClipmap Storer = new VoxelStorerClipmap
+                    {
+                        storageUints = storageUints,
+                        FragmentsBuffer = FragmentsBuffer,
+                        ClipMapCount = ClipMapCount,
+                        ClipMapCurrent = i,
+                        ClipMapResolution = ClipMapResolution,
+                        PerMapOffsetScale = PerMapOffsetScale,
+                        UpdatesPerFrame = UpdateMethods.SingleClipmap
+                    };
+                    foreach (var attr in data.Attributes)
+                    {
+                        attr.Attribute.CollectVoxelizationPasses(data.passList, Storer, BaseVoxelMatrix, ClipMapResolution, attr.Stage, attr.Output);
+                    }
+                }
+            }
+        }
+
+        Xenko.Rendering.ComputeEffect.ComputeEffectShader BufferToTextureColumns;
+        Xenko.Rendering.ComputeEffect.ComputeEffectShader ClearBuffer;
+
+        public void PostProcess(VoxelStorageContext storageContext, RenderDrawContext drawContext, ProcessedVoxelVolume data)
+        {
+            if (Math.Max(Math.Max(ClipMapResolution.X, ClipMapResolution.Y), ClipMapResolution.Z) < 32)
+                return;
+            if (FragmentsBuffer == null)
+                return;
+            var context = drawContext.RenderContext;
+            if (ClearBuffer == null)
+            {
+                ClearBuffer = new Xenko.Rendering.ComputeEffect.ComputeEffectShader(context) { ShaderSourceName = "ClearBuffer" };
+                //BufferToTexture = new Xenko.Rendering.ComputeEffect.ComputeEffectShader(Context) { ShaderSourceName = "BufferToTextureEffect" };
```

Comment:
Originally I intended to have it switch between two different compute shaders depending on whether every voxel was independent and so the buffer could be processed in any order, or if it needed to be processed in columns (the solidify attribute needs this); I suspected performance would be better not having to loop over columns unless necessary. 

However the column processing seemed to perform fine, and the other code path eroded over time - so I never got around to doing a proper performance comparison. I'll get it working again and see if it's worth keeping both ways, otherwise the answer is no :)